File: hard/23.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* l, ListNode* r) {
            return l->val > r->val;
        };
        std::priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> minHeap(cmp);

        for (ListNode* head : lists) 
            if (head != nullptr) 
                minHeap.push(head);
        
        ListNode dummy(0);
        ListNode* tail = &dummy;

        while (!minHeap.empty()) {
            ListNode* smallest = minHeap.top();
            minHeap.pop();

            tail->next = smallest;
            tail = tail->next;

            if (smallest->next != nullptr) 
                minHeap.push(smallest->next);
        }
        return dummy.next;
    }
};

File: hard/23.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from heapq import heappush, heappop

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = []

        for i in range(len(lists)):
            if lists[i]:
                heappush(min_heap, (lists[i].val, i, lists[i]))
        
        dummy = ListNode(0)
        curr = dummy

        while min_heap:
            val, i, node = heappop(min_heap)
            curr.next = node
            curr = curr.next

            if node.next:
                heappush(min_heap, (node.next.val, i, node.next))
            
        return dummy.next
